### PR TITLE
Paul K Unit 4-2

### PR DESCRIPTION
**Pull request by llpk79**

### CODE REVIEW

**Review by hughjafro — COMMENTED** [review]
**Overall:**  You did an awesome job overall and your classes are top-notch.    

**_Part 1 - Define Terms:_**
You correctly defined all terms, explanations show a particularly strong understanding and mastery of these fundamental Neural Network terms.

**_Part 2 - Perceptron Implementation:_**
Your implementation is particularly elegant and readable, well commented.

**_Part 3 - MLP Implementation:_**
Your implementation is particularly elegant and readable, and well commented. Your implementation shows a substantial increase in accuracy as it is trained. Your MLP implementation correctly initializes weights, takes in inputs, combines inputs and weights as a weighted sum, activates the weighted sum, and passes the activated values onto the next layer. Backpropagation implementation appropriately updates weights in reverse order and leads to performance improvements in the model.

**_Part 4 - Topic Modeling:_**
Your Keras implementation appropriately uses a sequential model with dense layers, and appropriate architecture with *at least* _one hidden layer, activation functions, appropriate loss function, and accuracy metric_. Model’s verbose output shows significant improvement from first to last epoch during training. You showed your work in hyperparameter tuning two features and reported the CV accuracy of each parameter combination that is being tested.
